io/src/main/scala/fs2/io/io.scala

```diff
@@ -146,9 +156,25 @@ package object io {
 
     Stream.resource(mkOutput).flatMap {
       case (os, is) =>
-        val write = f(os) *> Sync[F].delay(os.close())
-        val read = readInputStream(is.pure[F], chunkSize, blocker, closeAfterUse = false)
-        read.concurrently(Stream.eval(write))
+        Stream.eval(Deferred[F, Option[Throwable]]).flatMap { err =>
+          // We need to close the output stream regardless of how `f` finishes
```

Comment:
Good find. I think I never ran into this because of the way I used the combinator in practice.